core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -198,9 +218,35 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    timeIndex.maybeAppend(largestTimestamp, lastValidOffset + 1)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadLargestTimestamp(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will retrun (-1, baseOffset)
+    val lastTimeIndexEntry = timeIndex.lastEntry
+    if (readToLogEnd) {
+      val offsetPosition = index.lookup(lastTimeIndexEntry.offset)
+      // Scan the rest of the messages to see if there is a larger timestamp after the last time index entry.
+      maxTimestampSoFar = math.max(lastTimeIndexEntry.timestamp, log.largestTimestampAfter(offsetPosition.position))
```

Comment:
Not sure why we need max here since we should always be able to find a timestamp >= lastTimeIndexEntry.timestamp.